Backend/Remora.Discord.Gateway/Transport/WebSocketPayloadTransportService.cs

```diff
@@ -172,58 +196,58 @@ public async Task<Result> SendPayloadAsync(IPayload payload, CancellationToken c
     /// <inheritdoc />
     public async Task<Result<IPayload>> ReceivePayloadAsync(CancellationToken ct = default)
     {
-        if (_clientWebSocket is null)
+        if (_clientWebSocket?.State is not WebSocketState.Open)
         {
             return new InvalidOperationError("The transport service is not connected.");
         }
 
-        if (_clientWebSocket.State != WebSocketState.Open)
-        {
-            return new InvalidOperationError("The socket was not open.");
-        }
-
-        await using var memoryStream = new MemoryStream();
-
-        var buffer = ArrayPool<byte>.Shared.Rent(4096);
+        var semaphoreEntered = false;
 
         try
         {
-            WebSocketReceiveResult result;
+            semaphoreEntered = await _receiveSemaphore.WaitAsync(100, ct);
+            if (!semaphoreEntered)
+            {
+                return new OperationCanceledException("Could not enter semaphore.");
+            }
 
+            await using var ms = _memoryStreamManager.GetStream();
```

Comment:
I think Jax would advise against abbreviated variable names like this 😄

Reply:
You're probably right... given the existing name was `memoryStream`, I'll change it back lol